### PR TITLE
 Fix process_array_expr to take into account pointer offset

### PR DESCRIPTION
**Pull request by tautschnig**
Includes code cleanup to remove an unnecessary use of process_array_expr.

### CODE REVIEW

**Review by peterschrammel — COMMENTED** [review]
Nice piece of work. 
Is there a test case for array_replace? I don't see how it is called in your test case.

**Comment by tautschnig** [comment]
> Is there a test case for array_replace? I don't see how it is called in your test case.

It is called from (our models of) the C library functions, including `memcpy`. Is that sufficient?

**Review by peterschrammel — APPROVED** [review]
Found it. I searched in the wrong file.

File: src/goto-symex/symex_clean_expr.cpp [hunk 1]
```diff
@@ -137,10 +60,57 @@ void goto_symext::process_array_expr(exprt &expr)
...
+        expr =
+          byte_extract_exprt(
+            byte_extract_id(),
+            expr,
+            from_integer(0, index_type()),
+            array_typet(char_type(), array_size));
+      }
+
+      // given an array type T[S] and a byte offset o, build a new type
+      // T[S-(o/sizeof(T))]
```
Comment (chrisr-diffblue) [inline src/goto-symex/symex_clean_expr.cpp]: I may have been particularly slow reading this comment (apologies if so) - but it wasn't immediately obvious to me that this type conversion (particularly the `o/sizeof(T)`) is to do the conversion from byte offset -> index. Maybe it's worth expanding the comment very slightly?

File: src/ansi-c/ansi_c_internal_additions.cpp [hunk 2]
```diff
@@ -221,8 +221,6 @@ void ansi_c_internal_additions(std::string &code)
     "float __CPROVER_fabsf(float x);\n"
 
     // arrays
-    // NOLINTNEXTLINE(whitespace/line_length)
-    "__CPROVER_bool __CPROVER_array_equal(const void *array1, const void *array2);\n"
     // overwrite all of *dest (possibly using nondet values), even
```
Comment (martin-cs) [inline src/ansi-c/ansi_c_internal_additions.cpp]: I like the idea of removing special case passes, particularly ones that are sharing unaware BUT it seems like array_copy only captures some of the functionality of array_equals.  With array_equals I can build two arrays in different ways and then test their equality.

File: src/goto-symex/symex_clean_expr.cpp [hunk 3]
```diff
@@ -94,41 +30,28 @@ void goto_symext::process_array_expr(exprt &expr)
...
   }
-  else if(expr.id()==ID_byte_extract_little_endian ||
-          expr.id()==ID_byte_extract_big_endian)
-  {
-    // pick the root object
-    exprt tmp=to_byte_extract_expr(expr).op();
-    expr.swap(tmp);
-    process_array_expr(expr);
-  }
   else if(expr.id()==ID_symbol &&
```
Comment (martin-cs) [inline src/goto-symex/symex_clean_expr.cpp:55]: There seem to be quite a few cases that have been refactored here that, to my eyes, are not covered by the test case added.

**Review by martin-cs — CHANGES_REQUESTED** [review]
Am concerned about the change in functionality but otherwise OK.

**Comment by tautschnig** [comment]
@martin-cs Thank you for getting me back on track, I'm not sure what I was doing when crafting ad728880fa55. I'll rework.

**Comment by martin-cs** [comment]
@tautschnig As I said about another commit -- about half of it is "hell yes we should have done that years ago"...

**Comment by tautschnig** [comment]
I have reworked all the array_equal stuff and now also included a specific regression test (rather than the more implicit ones before).

**Comment by tautschnig** [comment]
@martin-cs Would you mind taking another look?

**Review by martin-cs — APPROVED** [review]
Looks much better.  I would still prefer some more tests for the second commit but I won't block the merge because of it.

**Comment by tautschnig** [comment]
Just to note that the request for more tests is not being ignored: I'm trying to source them, but please bear with me as that might take some time.

**Comment by martin-cs** [comment]
Sure; it wasn't supposed to be a blocking request more a "I'd feel more confident about this if...".

**Pull request merged**